### PR TITLE
If password is empty, do not assume Plain Auth

### PR DESCRIPTION
**Pull request by rudderfree**
Pull request for #1382  

Assume the mail surver is not supporting AUTH

Signed-off-by: Amit Kumar Chaudhary <a.rudder.free@gmail.com>

### CODE REVIEW

**Comment by akramtexas** [comment]
@rudderfree I see a build-check (actually, two) has (have) failed. I've traced the build failure (for the `master-verify` job) to the following error:
```
08:55:10 GO111MODULE=on go vet ./...
08:55:22 gofmt -l .
08:55:23 internal/support/notifications/sending_service.go
08:55:23 [ "`gofmt -l .`" = "" ]
08:55:23 make: *** [test] Error 1
```

**Comment by rudderfree** [comment]
found the error :)

-       if (smtp.Password != "") {
+       if smtp.Password != "" {

**Comment by akramtexas** [comment]
Recheck

File: internal/support/notifications/sending_service.go [hunk 1]
```diff
@@ -84,15 +84,21 @@ func persistTransmission(tr models.TransmissionRecord, n models.Notification, c
...
 	buf := bytes.NewBufferString("Subject: " + smtp.Subject + "\r\n")
 	// required CRLF at ends of lines and CRLF between header and body for SMTP RFC 822 style email
 	buf.WriteString("\r\n")
 	buf.WriteString(message)
-	err := mail.SendMail(smtp.Host+":"+strconv.Itoa(smtp.Port),
-		mail.PlainAuth("", smtp.Sender, smtp.Password, smtp.Host),
-		smtp.Sender, addressees, []byte(buf.String()))
+	if (smtp.Password != "") {
+		err = mail.SendMail(smtp.Host+":"+strconv.Itoa(smtp.Port),
+			mail.PlainAuth("", smtp.Sender, smtp.Password, smtp.Host),
```
[the remark this answers is not in the document]
Comment (AnthonyMBonafide) [reply internal/support/notifications/sending_service.go]: Also, the CI jobs are failing due to the `gofmt` tool returning an error due to formatting issues. You can run `gofmt -l .` to see the packages which are causing the errors and you can run `gofmt -w .` to have the go formatting tool correct the formatting issues.

**Comment by akramtexas** [comment]
> @rudderfree I see a build-check (actually, two) has (have) failed. I've traced the build failure (for the `master-verify` job) to the following error:
> 
> ```
> 08:55:10 GO111MODULE=on go vet ./...
> 08:55:22 gofmt -l .
> 08:55:23 internal/support/notifications/sending_service.go
> 08:55:23 [ "`gofmt -l .`" = "" ]
> 08:55:23 make: *** [test] Error 1
> ```
> 
> While I can't say _definitively_ what is causing that error, I _can_ suggest that you try running the following command at your command-line, positioned at your project root level directory:
> 
> * `gofmt -w .`
> * This will format your code base, and make it conform to the Go standard.
> * Rebuild your feature branch, do `make test`, and push to the remote repo.

@rudderfree Was this helpful? (I notice that the build-checks are still failing...)

File: internal/support/notifications/sending_service.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rudderfree) [reply internal/support/notifications/sending_service.go]: Thanks, yes, the fix above, removing the brackets around the if condition [`if (smtp.Password != nil) {` ] was made by `gofmt -w .` itself, I am pushing the patch up with this change added, after reverting the current one.

Also trying out the smtp server in a docker container, and planning to test the changes with it before pushing.

[the remark this answers is not in the document]
Comment (rudderfree) [reply internal/support/notifications/sending_service.go]: Not yet tested with the postfix server for the authenticated case, resolving signed certificate issues. But it passes the no auth case with my machine's sendmail.

 `make test` and ` gofmt -l . ` doesn't show any errors. I think it should be ok to merge, but at the same time I am equally inclined to wait till I test it with the authenticated server also, which I will report back on as soon as I get it running properly.

**Comment by rudderfree** [comment]
> > @rudderfree I see a build-check (actually, two) has (have) failed. I've traced the build failure (for the `master-verify` job) to the following error:
> > ```
> > 08:55:10 GO111MODULE=on go vet ./...
> > 08:55:22 gofmt -l .
> > 08:55:23 internal/support/notifications/sending_service.go
> > 08:55:23 [ "`gofmt -l .`" = "" ]
> > 08:55:23 make: *** [test] Error 1
> > ```
> > 
> > 
> > While I can't say _definitively_ what is causing that error, I _can_ suggest that you try running the following command at your command-line, positioned at your project root level directory:
> > 
> > * `gofmt -w .`
> > * This will format your code base, and make it conform to the Go standard.
> > * Rebuild your feature branch, do `make test`, and push to the remote repo.
> 
> @rudderfree Was this helpful? (I notice that the build-checks are still failing...)

@akramtexas yes it was :)

**Comment by akramtexas** [comment]
@rudderfree Please update your feature branch. Also be sure to use Git to `squash` commits (down to a single commit).

**Comment by AnthonyMBonafide** [comment]
@rudderfree, with your last push you added a commit that was not signed which results in a failed `DCO` CI job. You can fix this by squashing all of your commits into the earliest one, which is signed.

**Comment by rudderfree** [comment]
I have pushed the squashed commit, there is a good chance something might still need some attention, if that is the case, I will be getting back to it on Saturday, next two days are packed.

**Comment by akramtexas** [comment]
> I have pushed the squashed commit, there is a good chance something might still need some attention, if that is the case, I will be getting back to it on Saturday, next two days are packed.

Thanks @rudderfree Also, please update your feature branch (so that it is no longer out-of-date with the base branch).

**Comment by rudderfree** [comment]
I pressed the **update branch** button, presented by this PR page. As a result of which my commit is no longer the top commit, and an extra merge commit is present on top. Is it ok? or is it required that my commit should have been the top commit? which means applying the changes again after re-basing the branch (rudderfree/edgex-go) with upstream (edgexfoundry/edgex-go).

This action has indeed pulled in the changes from upstream to my branch. The only thing I am not sure is if this is sufficient.

**Comment by AnthonyMBonafide** [comment]
> I pressed the **update branch** button, presented by this PR page. As a result of which my commit is no longer the top commit, and an extra merge commit is present on top. Is it ok? or is it required that my commit should have been the top commit? which means applying the changes again after re-basing the branch (rudderfree/edgex-go) with upstream (edgexfoundry/edgex-go).

For the all of the EdgeX repos, we prefer to avoid merge commits by re-basing, and squashing commits down to one commit to have a cleaner history.

**Comment by AnthonyMBonafide** [comment]
@rudderfree, did you intentionally close this PR?

**Comment by rudderfree** [comment]
no, I just reset my master branch up to the commit that was below my commit, so probably that caused it

**Comment by rudderfree** [comment]
re-opening, it was found closed for unknown reason.

re-based the branch, and everything looks in order now.

**Comment by akramtexas** [comment]
Recheck

**Comment by akramtexas** [comment]
@rudderfree For some reason, one build-check (for `DCO`) is failing. When squashing commits and/or rebasing, one line (such as the following), should be in the commit-editor, prior to pushing to remote repo: `Signed-off-by: Amit Kumar Chaudhary a.rudder.free@gmail.com`

**Comment by AnthonyMBonafide** [comment]
@rudderfree , any luck with testing with username and password authentication?

**Comment by rudderfree** [comment]
> @rudderfree , any luck with testing with username and password authentication?

not yet, certificates are not working for me, smtp is returning with 
`errorx509: certificate signed by unknown authority`, looking into it.

**Comment by rudderfree** [comment]
> @rudderfree For some reason, one build-check (for `DCO`) is failing. When squashing commits and/or rebasing, one line (such as the following), should be in the commit-editor, prior to pushing to remote repo: `Signed-off-by: Amit Kumar Chaudhary a.rudder.free@gmail.com`

All the commits seems to be signed as far as I can see, there is no additional re-base commit. Is it possible for you to cut paste the error.

**Comment by rudderfree** [comment]
> > @rudderfree , any luck with testing with username and password authentication?
> 
> not yet, certificates are not working for me, smtp is returning with
> `errorx509: certificate signed by unknown authority`, looking into it.
postfix container seems to be running fine, by running fine I mean I can telnet to it.

**Comment by akramtexas** [comment]
@rudderfree This is the `DCO` check I was referring to (To be precise, it's not `failing`; for some reason, it never gets to run...)
DCO check

**Comment by AnthonyMBonafide** [comment]
> not yet, certificates are not working for me, smtp is returning with
> `errorx509: certificate signed by unknown authority`

This is the situation I mentioned in my previous comment. As of now, we can only connect and leverage an SMTP server which is configured to not use authentication(or blank username with a password), and configured with a valid non self-signed cert. We can get over these limitations by doing the following:

1. Allow the user to provide a username and password which will be sent to the SMTP server during the initial connection via the SMTP client. Right now we are passing along the password to the SMTP client but not a username. Therefore we can only authenticate if an empty/blank username is valid.

1. Allow the SMTP client the ability to ignore "invalid" certs, by invalid I mean self-signed certs. For example, when using the Go HTTP client you can do the following to allow the client to use self signed certs:
```
client := &http.Client{
                Transport: &http.Transport{
                       TLSClientConfig: &tls.Config{
                               InsecureSkipVerify: true,
                       },
                },
        }
```
My thought is that if there is something similar we can do for the SMTP client then we can make it configurable(Off by default), so that the user has the ability to connect to an SMTP with a self signed-cert if they choose, also we can easily test.

Does all of that make sense?

**Comment by rudderfree** [comment]
> > not yet, certificates are not working for me, smtp is returning with
> > `errorx509: certificate signed by unknown authority`
> 
> This is the situation I mentioned in my previous comment. As of now, we can only connect and leverage a SMTP server which is configured to not use authentication(or blank username with a password), and configured with a valid non self-signed cert. We can get over these limitations by doing the following:
> 
>     1. Allow the user to provide a username and password which will be sent to the SMTP server during the initial connection via the SMTP client. Right now we are passing along the password to the SMTP client but not a username. Therefore we can only authenticate if an empty/blank username is valid.
> 
>     2. Allow the SMTP client the ability to ignore "invalid" certs, by invalid I mean self-signed certs. For example, when using the Go HTTP client you can do the following to allow the client to use self signed certs:
> 
> 
> ```
> client := &http.Client{
>                 Transport: &http.Transport{
>                        TLSClientConfig: &tls.Config{
>                                InsecureSkipVerify: true,
>                        },
>                 },
>         }
> ```
> 
> My thought is that if there is something similar we can do for the SMTP client then we can make it configurable(Off by default), so that the user has the ability to connect to an SMTP with a self signed-cert if they choose, also we can easily test.
> 
> Does all of that make sense?

To some degree, yes, let me look at it more closely. Probably opening another issue for this would be a good idea.

**Comment by rudderfree** [comment]
> > > not yet, certificates are not working for me, smtp is returning with
> > > `errorx509: certificate signed by unknown authority`
> > 
> > 
> > This is the situation I mentioned in my previous comment. As of now, we can only connect and leverage a SMTP server which is configured to not use authentication(or blank username with a password), and configured with a valid non self-signed cert. We can get over these limitations by doing the following:
> > ```
> > 1. Allow the user to provide a username and password which will be sent to the SMTP server during the initial connection via the SMTP client. Right now we are passing along the password to the SMTP client but not a username. Therefore we can only authenticate if an empty/blank username is valid.
> > 
> > 2. Allow the SMTP client the ability to ignore "invalid" certs, by invalid I mean self-signed certs. For example, when using the Go HTTP client you can do the following to allow the client to use self signed certs:
> > ```
> > 
> > 
> > ```
> > client := &http.Client{
> >                 Transport: &http.Transport{
> >                        TLSClientConfig: &tls.Config{
> >                                InsecureSkipVerify: true,
> >                        },
> >                 },
> >         }
> > ```
> > 
> > 
> > My thought is that if there is something similar we can do for the SMTP client then we can make it configurable(Off by default), so that the user has the ability to connect to an SMTP with a self signed-cert if they choose, also we can easily test.
> > Does all of that make sense?
> 
> To some degree, yes, let me look at it more closely. Probably opening another issue for this would be a good idea.

Yes, got your point.

**Comment by rudderfree** [comment]
opps, closed the PR, while canceling the comment. (re-opened)

**Comment by rudderfree** [comment]
> > > > not yet, certificates are not working for me, smtp is returning with
> > > > `errorx509: certificate signed by unknown authority`
> > > 
> > > 
> > > This is the situation I mentioned in my previous comment. As of now, we can only connect and leverage a SMTP server which is configured to not use authentication(or blank username with a password), and configured with a valid non self-signed cert. We can get over these limitations by doing the following:
> > > ```
> > > 1. Allow the user to provide a username and password which will be sent to the SMTP server during the initial connection via the SMTP client. Right now we are passing along the password to the SMTP client but not a username. Therefore we can only authenticate if an empty/blank username is valid.
> > > 
> > > 2. Allow the SMTP client the ability to ignore "invalid" certs, by invalid I mean self-signed certs. For example, when using the Go HTTP client you can do the following to allow the client to use self signed certs:
> > > ```
> > > 
> > > 
> > > ```
> > > client := &http.Client{
> > >                 Transport: &http.Transport{
> > >                        TLSClientConfig: &tls.Config{
> > >                                InsecureSkipVerify: true,
> > >                        },
> > >                 },
> > >         }
> > > ```
> > > 
> > > 
> > > My thought is that if there is something similar we can do for the SMTP client then we can make it configurable(Off by default), so that the user has the ability to connect to an SMTP with a self signed-cert if they choose, also we can easily test.
> > > Does all of that make sense?
> > 
> > 
> > To some degree, yes, let me look at it more closely. Probably opening another issue for this would be a good idea.
> 
> Yes, got your point.

It passes the test after adding` config.InsecureSkipVerify = true` to 
`func SendMail(addr string, a Auth, from string, to []string, msg []byte) error {` from
go/src/net/smtp/smtp.go

```
        if ok, _ := c.Extension("STARTTLS"); ok {
                config := &tls.Config{ServerName: c.serverName}
                config.InsecureSkipVerify = true
                if testHookStartTLS != nil {
```

**Comment by rudderfree** [comment]
Moved the conversation about username, password authenticated server to #1504

**Comment by rudderfree** [comment]
I am planning to combine this change with SkipVerify (#1504) change and create a new PR, I have looked up, and DCO have been reported to hang even when everything seems alright.

**Comment by akramtexas** [comment]
> I am planning to combine this change with SkipVerify (#1504) change and create a new PR, I have looked up, and DCO have been reported to hang even when everything seems alright.

@rudderfree If you're creating a new branch for the other, proposed PR (#1504), please spend some time determining how best to bring over all (relevant) changes from this current PR (#1469) to incorporate into the other PR. Please continue to work with @AnthonyMBonafide and pull me in as needed.

**Comment by AnthonyMBonafide** [comment]
@rudderfree , I am assuming you want to deal with the authentication issues in a different PR and branch and would like this to be reviewed and merged separate from issue #1504. Is that correct?

If so, @akramtexas do you find this acceptable?  

Also, please update this branch with the changes that have been merged into `master` and this PR can be re-evaluated.

**Comment by akramtexas** [comment]
> @rudderfree , I am assuming you want to deal with the authentication issues in a different PR and branch and would like this to be reviewed and merged separate from issue #1504. Is that correct?
> 
> If so, @akramtexas do you find this acceptable?
> 
> Also, please update this branch with the changes that have been merged into `master` and this PR can be re-evaluated.

- Nicely articulated, @AnthonyMBonafide 👍 
- @rudderfree Please proceed and deal with the (larger) authentication issues in that other PR. Meanwhile, yes, continue using this existing PR for fixing the issue of EdgeX passing an empty string as a password rather than `nil`, thanks.

**Comment by AnthonyMBonafide** [comment]
@rudderfree , It seems like the DCO check is stuck in a bad state. Could you create a new source branch for this PR so that the CI job successfully kicks in? Sorry about that, sometimes this happens when the DCO check fails on earlier commits. Once that is fixed I will re-verify this PR, as it looks like it is in a good state.

**Comment by akramtexas** [comment]
> @rudderfree , It seems like the DCO check is stuck in a bad state. Could you create a new source branch for this PR so that the CI job successfully kicks in? Sorry about that, sometimes this happens when the DCO check fails on earlier commits. Once that is fixed I will re-verify this PR, as it looks like it is in a good state.
> 
> Also, update your branch with the latest changes from edgex-go master

@AnthonyMBonafide Excellent recommendation 👍 
@rudderfree Thanking you for your diligence. And once you've got the new branch and new PR going, looking forward to further contributions from you 👍

**Comment by rudderfree** [comment]
> @rudderfree , It seems like the DCO check is stuck in a bad state. Could you create a new source branch for this PR so that the CI job successfully kicks in? Sorry about that, sometimes this happens when the DCO check fails on earlier commits. Once that is fixed I will re-verify this PR, as it looks like it is in a good state.
Yes, I was thinking along the same lines

> Also, update your branch with the latest changes from edgex-go master
Will do.

**Comment by rudderfree** [comment]
> > Also, update your branch with the latest changes from edgex-go master
> 
> @AnthonyMBonafide Excellent recommendation +1
> @rudderfree Thanking you for your diligence. And once you've got the new branch and new PR going (to take care of the `DCO` check issue), looking forward to further contributions from you +1

Sure thing +1

**Comment by rudderfree** [comment]
Closing, to re-fork a clean master from upstream.

**Comment by akramtexas** [comment]
Excellent coordination on your diligence and working through the details, gents: @rudderfree and @AnthonyMBonafide 👍

**Pull request closed without merging**